File: callouts/python/extauthz/tests/jwt_auth_test.py
# Copyright 2025 Google LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from __future__ import print_function

import datetime
from http.server import BaseHTTPRequestHandler
from http.server import HTTPServer
import threading
import time
from typing import Iterator, Callable, Any, Mapping
import urllib.request
import ssl
from unittest.mock import patch

from envoy.service.auth.v3 import external_auth_pb2 as auth_pb2
from envoy.service.auth.v3 import external_auth_pb2_grpc as auth_pb2_grpc
from envoy.service.auth.v3 import attribute_context_pb2 as attr_pb2
from envoy.type.v3 import http_status_pb2
import grpc
import pytest

from extauthz.example.jwt_auth.service_callout_example import (
    JwtAuthServer as CalloutServerTest,
)
from extauthz.service.callout_server import CalloutServerAuth, _addr_to_str


class ServerSetupException(Exception):
    pass


class NoResponseError(Exception):
    pass


default_kwargs: dict = {
    'address': ('localhost', 8443),
    'plaintext_address': ('localhost', 8080),
    'health_check_address': ('localhost', 8000)
}
# Arguments for running a custom CalloutServer with testing parameters.
_local_test_args: dict = {
    "kwargs": default_kwargs,
    "test_class": CalloutServerTest
}


def get_plaintext_channel(server: CalloutServerAuth) -> grpc.Channel:
    """From a CalloutServer, obtain the plaintext address and create a grpc channel pointing to it.

    Args:
        server: Server to connect to.
    Returns:
        grpc.Channel: Open channel to the server.
    """
    addr = server.plaintext_address
    return grpc.insecure_channel(_addr_to_str(addr) if addr else '')


def wait_till_server(server_check: Callable[[], bool], timeout: int = 10):
    """Wait until the `server_check` function returns true.

    Used for blocking until the server reaches a given state.
    Times out after a given time.

    Args:
        server_check: Function to check.
        timeout: Wait time. Defaults to 10.
    """
    expiration = datetime.datetime.now() + datetime.timedelta(seconds=timeout)
    while not server_check() and datetime.datetime.now() < expiration:
        time.sleep(1)


def _start_server(server: CalloutServerAuth) -> threading.Thread:
    # Start the server in a background thread
    thread = threading.Thread(target=server.run)
    thread.daemon = True
    thread.start()
    # Wait for the server to start
    wait_till_server(lambda: getattr(server, '_setup', False))
    return thread


def _stop_server(server: CalloutServerAuth, thread: threading.Thread):
    # Stop the server
    server.shutdown()
    thread.join(timeout=5)


@pytest.fixture(scope='class', name='server')
def setup_server(request) -> Iterator[CalloutServerAuth]:
    """Set up basic CalloutServer.

    Takes in two optional pytest parameters.
    'kwargs': Arguments passed into the server constructor. 
      Default is the value of default_kwargs.
    'test_class': Class to use when constructing the server.
      Default is the base CalloutServer.

    Yields:
        Iterator[CalloutServerAuth]: The server to test with.
    """
    params: dict = request.param or {'kwargs': {}, 'test_class': None}
    kwargs: Mapping[str, Any] = default_kwargs | params['kwargs']
    # Either use the provided class or create a server using the default CalloutServer class.
    server = (params['test_class'] or CalloutServerAuth)(**kwargs)
    try:
        thread = _start_server(server)
        yield server
        _stop_server(server, thread)
    finally:
        del server


def make_request(stub: auth_pb2_grpc.AuthorizationStub, request: auth_pb2.CheckRequest) -> auth_pb2.CheckResponse:
    """Make a request to the server.

    Args:
        stub: The server stub.
        request: The CheckRequest to send.

    Returns: The CheckResponse returned from the server.
    """
    try:
        return stub.Check(request)
    except Exception as e:
        raise NoResponseError(f"Request failed: {e}")


def create_request_with_auth_header(auth_header: str) -> auth_pb2.CheckRequest:
    """Helper to create a request with Authorization header."""
    return auth_pb2.CheckRequest(
        attributes=attr_pb2.AttributeContext(
            request=attr_pb2.AttributeContext.Request(
                http=attr_pb2.AttributeContext.HttpRequest(
                    headers={'authorization': auth_header}
                )
            )
        )
    )


class TestJwtAuthServer(object):
    """Test server functionality for JWT authentication."""

    @pytest.mark.parametrize('server', [_local_test_args], indirect=True)
    def test_valid_jwt_token(self, server: CalloutServerTest) -> None:
        """Test that requests with valid JWT tokens are allowed."""
        # Mock the validate_jwt_token method to return a valid decoded token
        with patch.object(server, 'validate_jwt_token', return_value={
            'sub': '1234567890',
            'name': 'John Doe', 
            'admin': True,
            'iat': 1712173461,
            'exp': 2075658261
        }):
            with get_plaintext_channel(server) as channel:
                stub = auth_pb2_grpc.AuthorizationStub(channel)
                
                auth_header = "Bearer valid_token"
                request = create_request_with_auth_header(auth_header)
                response = make_request(stub, request)

                assert response.HasField('ok_response')
                # Verify decoded fields were added as headers
                ok_headers = {header.header.key: header.header.value for header in response.ok_response.headers}
                assert ok_headers.get('decoded-sub') == '1234567890'
                assert ok_headers.get('decoded-name') == 'John Doe'
                assert ok_headers.get('decoded-admin') == 'True'

    @pytest.mark.parametrize('server', [_local_test_args], indirect=True)
    def test_invalid_jwt_token(self, server: CalloutServerTest) -> None:
        """Test that requests with invalid JWT tokens are denied."""
        # Mock the validate_jwt_token method to return None (invalid token)
        with patch.object(server, 'validate_jwt_token', return_value=None):
            with get_plaintext_channel(server) as channel:
                stub = auth_pb2_grpc.AuthorizationStub(channel)
                
                auth_header = "Bearer invalid_token"
                request = create_request_with_auth_header(auth_header)
                response = make_request(stub, request)

                assert response.HasField('denied_response')
                assert response.denied_response.status.code == http_status_pb2.StatusCode.Unauthorized
                assert "Authorization token is invalid" in response.denied_response.body

    @pytest.mark.parametrize('server', [_local_test_args], indirect=True)
    def test_missing_jwt_token(self, server: CalloutServerTest) -> None:
        """Test that requests without JWT tokens are denied."""
        with get_plaintext_channel(server) as channel:
            stub = auth_pb2_grpc.AuthorizationStub(channel)
            
            # Create a request without the Authorization header
            request = auth_pb2.CheckRequest(
                attributes=attr_pb2.AttributeContext(
                    request=attr_pb2.AttributeContext.Request(
                        http=attr_pb2.AttributeContext.HttpRequest(
                            headers={}
                        )
                    )
                )
            )
            response = make_request(stub, request)

            assert response.HasField('denied_response')
            assert response.denied_response.status.code == http_status_pb2.StatusCode.Unauthorized
            assert "No Authorization token found" in response.denied_response.body

    @pytest.mark.parametrize('server', [_local_test_args], indirect=True)
    def test_basic_server_health_check(self, server: CalloutServerTest) -> None:
        """Test that the health check sub server returns the expected 200 code."""
        assert server.health_check_address is not None
        response = urllib.request.urlopen(
            f'http://{_addr_to_str(server.health_check_address)}')
        assert not response.read()
        assert response.getcode() == 200

File: callouts/go/extauthz/examples/jwt_auth/jwt_auth_test.go
// Copyright 2025 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package jwt_auth

import (
	"context"
	"os"
	"path/filepath"
	"testing"

	"github.com/dgrijalva/jwt-go"
	authv3 "github.com/envoyproxy/go-control-plane/envoy/service/auth/v3"
	typev3 "github.com/envoyproxy/go-control-plane/envoy/type/v3"
)

// getTestFilePath retorna o caminho absoluto para arquivos de teste
func getTestFilePath(filename string) string {
	return filepath.Join("..", "..", "ssl_creds", filename)
}

// generateTestJWTToken generates a JWT token for testing purposes.
func generateTestJWTToken(privateKey []byte, claims jwt.MapClaims) (string, error) {
	token := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
	privateKeyParsed, err := jwt.ParseRSAPrivateKeyFromPEM(privateKey)
	if err != nil {
		return "", err
	}
	return token.SignedString(privateKeyParsed)
}

// TestHandleCheck_ValidToken tests the handling of check requests with a valid JWT token.
func TestHandleCheck_ValidToken(t *testing.T) {
	// Load the test private key
	privateKey, err := os.ReadFile(getTestFilePath("localhost.key"))
	if err != nil {
		t.Fatalf("failed to load private key: %v", err)
	}

	// Create a test JWT token
	claims := jwt.MapClaims{
		"sub":  "1234567890",
		"name": "John Doe",
		"iat":  int64(1720020355),
		"exp":  int64(1820023955),
	}
	tokenString, err := generateTestJWTToken(privateKey, claims)
	if err != nil {
		t.Fatalf("failed to generate test JWT token: %v", err)
	}

	// Create an instance of JwtAuthServer with the test public key path
	service := NewJwtAuthServerWithKeyPath(getTestFilePath("publickey.pem"))

	req := &authv3.CheckRequest{
		Attributes: &authv3.AttributeContext{
			Request: &authv3.AttributeContext_Request{
				Http: &authv3.AttributeContext_HttpRequest{
					Headers: map[string]string{
						"authorization": "Bearer " + tokenString,
					},
				},
			},
		},
	}

	// Call the HandleCheck method
	response, err := service.HandleCheck(context.Background(), req)
	if err != nil {
		t.Fatalf("HandleCheck got err: %v", err)
	}

	// Check if the response is not nil
	if response == nil {
		t.Fatalf("HandleCheck(): got nil resp, want non-nil")
	}

	// Check if it's an OK response
	if response.GetOkResponse() == nil {
		t.Fatalf("Expected OK response for valid JWT")
	}

	// Prepare expected headers
	expectedHeaders := []struct {
		key   string
		value string
	}{
		{"decoded-sub", "1234567890"},
		{"decoded-name", "John Doe"},
		{"decoded-iat", "1720020355"},
		{"decoded-exp", "1820023955"},
	}

	// Check headers in the OK response
	headers := response.GetOkResponse().Headers
	for _, expected := range expectedHeaders {
		found := false
		for _, h := range headers {
			if h.Header.Key == expected.key && h.Header.Value == expected.value {
				found = true
				break
			}
		}
		if !found {
			t.Errorf("Expected header %s: %s not found", expected.key, expected.value)
		}
	}
}

// TestHandleCheck_InvalidToken tests the handling of check requests with an invalid JWT token.
func TestHandleCheck_InvalidToken(t *testing.T) {
	// Create an instance of JwtAuthServer with a test public key path
	service := NewJwtAuthServerWithKeyPath(getTestFilePath("publickey.pem"))
	// Override the public key to be invalid
	service.PublicKey = []byte("invalidpublickey")

	req := &authv3.CheckRequest{
		Attributes: &authv3.AttributeContext{
			Request: &authv3.AttributeContext_Request{
				Http: &authv3.AttributeContext_HttpRequest{
					Headers: map[string]string{
						"authorization": "Bearer invalidtoken",
					},
				},
			},
		},
	}

	// Call the HandleCheck method
	response, err := service.HandleCheck(context.Background(), req)
	if err != nil {
		t.Fatalf("HandleCheck got err: %v", err)
	}

	// Check if it's a denied response
	if response.GetDeniedResponse() == nil {
		t.Fatalf("Expected denied response for invalid JWT")
	}

	if response.GetDeniedResponse().Status.Code != typev3.StatusCode_Unauthorized {
		t.Errorf("Expected status code Unauthorized, got %v", response.GetDeniedResponse().Status.Code)
	}

	body := response.GetDeniedResponse().Body
	expectedBody := "Authorization token is invalid."
	if body != expectedBody {
		t.Errorf("Expected body %s, got %s", expectedBody, body)
	}
}

// TestHandleCheck_MissingToken tests the handling of check requests without a JWT token.
func TestHandleCheck_MissingToken(t *testing.T) {
	// Use absolute path for the test
	service := NewJwtAuthServerWithKeyPath(getTestFilePath("publickey.pem"))

	req := &authv3.CheckRequest{
		Attributes: &authv3.AttributeContext{
			Request: &authv3.AttributeContext_Request{
				Http: &authv3.AttributeContext_HttpRequest{
					Headers: map[string]string{},
				},
			},
		},
	}

	response, err := service.HandleCheck(context.Background(), req)
	if err != nil {
		t.Fatalf("HandleCheck got err: %v", err)
	}

	if response.GetDeniedResponse() == nil {
		t.Fatalf("Expected denied response for missing JWT")
	}

	if response.GetDeniedResponse().Status.Code != typev3.StatusCode_Unauthorized {
		t.Errorf("Expected status code Unauthorized, got %v", response.GetDeniedResponse().Status.Code)
	}

	body := response.GetDeniedResponse().Body
	expectedBody := "No Authorization token found."
	if body != expectedBody {
		t.Errorf("Expected body %s, got %s", expectedBody, body)
	}
}
